src/modules/silentpayments/main_impl.h

```diff
@@ -0,0 +1,713 @@
+    for (i = 0; i < n_recipients; i++) {
+        if ((i == 0) || (secp256k1_ec_pubkey_cmp(ctx, &current_scan_pubkey, &recipients[i]->scan_pubkey) != 0)) {
+            /* If we are on a different scan pubkey, its time to recreate the shared secret and reset k to 0.
+             * It's very unlikely the scan public key is invalid by this point, since this means the caller would
+             * have created the _silentpayments_recipient object incorrectly, but just to be sure we still check that
+             * the public key is valid.
+             */
+            secp256k1_ge pk;
+            if (!secp256k1_pubkey_load(ctx, &pk, &recipients[i]->scan_pubkey)) {
+                secp256k1_scalar_clear(&a_sum_scalar);
+                /* Leaking this value would break indistiguishability of the transaction, so clear it. */
+                secp256k1_memclear(&shared_secret, sizeof(shared_secret));
+                return 0;
+            }
+            secp256k1_silentpayments_create_shared_secret(ctx, shared_secret, &a_sum_scalar, &pk);
+            k = 0;
+        }
+        if (!secp256k1_silentpayments_create_output_pubkey(ctx, generated_outputs[recipients[i]->index], shared_secret, &recipients[i]->spend_pubkey, k)) {
+            secp256k1_scalar_clear(&a_sum_scalar);
+            secp256k1_memclear(&shared_secret, sizeof(shared_secret));
+            return 0;
+        }
+        k++;
+        current_scan_pubkey = recipients[i]->scan_pubkey;
+    }
+    secp256k1_scalar_clear(&a_sum_scalar);
+    secp256k1_memclear(&shared_secret, sizeof(shared_secret));
+    return 1;
+}
+
+/** Set hash state to the BIP340 tagged hash midstate for "BIP0352/Label". */
+static void secp256k1_silentpayments_sha256_init_label(secp256k1_sha256* hash) {
+    secp256k1_sha256_initialize(hash);
+    hash->s[0] = 0x26b95d63ul;
+    hash->s[1] = 0x8bf1b740ul;
+    hash->s[2] = 0x10a5986ful;
+    hash->s[3] = 0x06a387a5ul;
+    hash->s[4] = 0x2d1c1c30ul;
+    hash->s[5] = 0xd035951aul;
+    hash->s[6] = 0x2d7f0f96ul;
+    hash->s[7] = 0x29e3e0dbul;
+
+    hash->bytes = 64;
+}
+
+int secp256k1_silentpayments_recipient_create_label(const secp256k1_context *ctx, secp256k1_pubkey *label, unsigned char *label_tweak32, const unsigned char *recipient_scan_key32, const uint32_t m) {
+    secp256k1_sha256 hash;
+    unsigned char m_serialized[4];
+
+    /* Sanity check inputs. */
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(label != NULL);
+    ARG_CHECK(label_tweak32 != NULL);
+    ARG_CHECK(recipient_scan_key32 != NULL);
+
+    /* Compute label_tweak = hash(ser_256(b_scan) || ser_32(m))  [sha256 with tag "BIP0352/Label"] */
+    secp256k1_silentpayments_sha256_init_label(&hash);
+    secp256k1_sha256_write(&hash, recipient_scan_key32, 32);
+    secp256k1_write_be32(m_serialized, m);
+    secp256k1_sha256_write(&hash, m_serialized, sizeof(m_serialized));
+    secp256k1_sha256_finalize(&hash, label_tweak32);
+
```

Comment:
nit (in `int secp256k1_silentpayments_recipient_create_label`):
```suggestion
secp256k1_memclear(m_serialized, sizeof(m_serialized));
secp256k1_sha256_clear(&hash);
```

Reply:
Done.